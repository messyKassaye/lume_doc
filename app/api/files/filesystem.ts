import fs from 'fs';
import path from 'path';
import { Readable } from 'stream';

import ID from 'shared/uniqueID';
import asyncFS from 'api/utils/async-fs';
import { tenants } from 'api/tenants/tenantContext';
import { testingTenants } from 'api/utils/testingTenants';
import { promisify } from 'util';

import { FileType } from '../../shared/types/fileType';

export type FilePath = string;
export type pathFunction = (fileName?: string) => FilePath;
const LOGS_DIR = './log';

const uploadsPath: pathFunction = (fileName: string = ''): FilePath =>
  path.join(tenants.current().uploadedDocuments, fileName);

const attachmentsPath: pathFunction = (fileName: string = ''): FilePath =>
  path.join(tenants.current().attachments, fileName);

const customUploadsPath: pathFunction = (fileName: string = ''): FilePath =>
  path.join(tenants.current().customUploads, fileName);

const temporalFilesPath: pathFunction = (fileName: string = ''): FilePath =>
  path.join(tenants.current().temporalFiles, fileName);

const activityLogPath: pathFunction = (): FilePath =>
  path.join(LOGS_DIR, `${tenants.current().name}_activity.log`);

async function deleteFile(file: FilePath) {
  return new Promise((resolve, reject) => {
    fs.unlink(file, err => {
      if (err && err.code !== 'ENOENT') {
        reject(err);
      }
      resolve();
    });
  });
}

async function deleteFiles(files: FilePath[]) {
  return Promise.all(files.map(async file => deleteFile(file)));
}

const createDirIfNotExists = async (dirPath: string) => {
  try {
    await asyncFS.mkdir(dirPath);
  } catch (e) {
    if (!e.message.match(/file already exists/)) {
      throw e;
    }
  }
};

const generateUploadsPath = async (subPath: string) => {
  if (subPath) {
    await createDirIfNotExists(`${__dirname}/specs/uploads/${subPath}`);
    await createDirIfNotExists(`${__dirname}/specs/customUploads/${subPath}`);
  }
  return {
    uploadedDocuments: `${__dirname}/specs/uploads/${subPath}`,
    attachments: `${__dirname}/specs/uploads/${subPath}`,
    customUploads: `${__dirname}/specs/customUploads/${subPath}`,
    temporalFiles: `${__dirname}/specs/uploads/${subPath}`,
  };
};

const setupTestUploadedPaths = async (subFolder: string = '') => {
  testingTenants.changeCurrentTenant(await generateUploadsPath(subFolder));
};

const deleteUploadedFiles = async (files: FileType[]) =>
  deleteFiles(
    files
      .filter(f => f.filename)
      .map(({ filename = '', type }) => {
        if (type === 'custom') {
          return customUploadsPath(filename);
        }
        return uploadsPath(filename);
      })
  );

const writeFile = promisify(fs.writeFile);
const appendFile = promisify(fs.appendFile);

const fileExists = async (filePath: FilePath): Promise<boolean> =>
  new Promise((resolve, reject) => {
    fs.access(filePath, err => {
      if (err === null) {
        resolve(true);
      }
      if (err?.code === 'ENOENT') {
        resolve(false);
      }
      if (err) {
        reject(err);
      }
    });
  });

const generateFileName = ({ originalname = '' }: FileType) =>
  Date.now() + ID() + path.extname(originalname);

/**
 * Create a file from a read stream and save it to one of uwazi filesystem paths
 * @param destination by default this will be uploadsPaths,
 * if you want another one you can pass filesystem destinatations
 * e.g. attachmentsPath()
 *
 */
const fileFromReadStream = async (
  fileName: FilePath,
  readStream: Readable,
  destination: string | undefined = undefined
): Promise<FilePath> =>
  new Promise((resolve, reject) => {
    const filePath = path.join(destination || uploadsPath(), fileName);
    const writeStream = fs.createWriteStream(filePath);
    readStream
      .pipe(writeStream)
      .on('finish', () => resolve(filePath))
      .on('error', reject);
  });

const streamToString = async (stream: Readable): Promise<string> =>
  new Promise((resolve, reject) => {
    const chunks: any[] = [];
    stream.on('data', (chunk: any) => chunks.push(chunk));
    stream.on('error', reject);
    stream.on('end', () => resolve(Buffer.concat(chunks).toString('utf8')));
  });

const getFileContent = async (fileName: FilePath): Promise<string> =>
  asyncFS.readFile(uploadsPath(fileName), 'utf8');

export {
  setupTestUploadedPaths,
  deleteUploadedFiles,
  createDirIfNotExists,
  deleteFiles,
  deleteFile,
  generateFileName,
  fileFromReadStream,
  fileExists,
  streamToString,
  getFileContent,
  customUploadsPath,
  uploadsPath,
  temporalFilesPath,
  attachmentsPath,
  activityLogPath,
  writeFile,
  appendFile,
};
