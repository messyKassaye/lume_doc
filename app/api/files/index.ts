export * from './filesystem';
export { PDF } from './PDF';
export { files } from './files';
export { uploadMiddleware } from './uploadMiddleware';
