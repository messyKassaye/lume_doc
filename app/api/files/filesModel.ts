import mongoose from 'mongoose';
import date from 'api/utils/date.js';

import { instanceModel } from 'api/odm';
import { FileType } from '../../shared/types/fileType';

const propsWithDBSpecifics = {
  creationDate: { type: Number, default: date.currentUTC },
  fullText: { type: mongoose.Schema.Types.Mixed, select: false },
  pdfInfo: { type: mongoose.Schema.Types.Mixed, select: false },
  entity: { type: String, index: true },
  type: { type: String, index: true },
  filename: { type: String, index: true },
};

const mongoSchema = new mongoose.Schema(propsWithDBSpecifics, {
  emitIndexErrors: true,
  strict: false,
});

const Model = instanceModel<FileType>('files', mongoSchema);

export default Model;
