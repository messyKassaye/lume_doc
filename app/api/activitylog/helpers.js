import { allLanguages } from 'shared/languagesList';
import { typeParsers } from 'api/activitylog/migrationsParser';
import templates from 'api/templates/templates';
import entities from 'api/entities/entities';
import { files } from 'api/files';

export const formatLanguage = langKey => {
  const lang = allLanguages.find(({ key }) => key === langKey);
  return lang ? `${lang.label} (${lang.key})` : langKey;
};

export const formatDataLanguage = data => formatLanguage(data.key);

export const translationsName = data => {
  const [context] = data.contexts;
  return data.contexts.length === 1
    ? `in ${context.label} (${context.id})`
    : 'in multiple contexts';
};

export const nameFunc = data => `${data.label} (${data.key})`;

export const migrationLog = log => {
  const data = JSON.parse(log.body);
  return typeParsers[data.type] ? typeParsers[data.type](data) : { action: 'RAW' };
};

export const templateName = data =>
  data.templateData ? `${data.templateData.name} (${data._id})` : data._id;

export const loadEntityFromPublicForm = async data => {
  const entity = JSON.parse(data.entity);
  const templateData = await templates.getById(entity.template);
  return { ...data, templateData, title: entity.title };
};

export const loadTemplate = async data => {
  const templateData = await templates.getById(data.template || data._id);
  return { ...data, templateData };
};

export const loadEntity = async data => {
  const _id = data.entityId || data._id;
  const sharedId = data.sharedId || data.entity;
  const query = { ...(_id && { _id }), ...(sharedId && { sharedId }) };
  const [entity] = await entities.get(query);
  return { ...data, entity, title: entity ? entity.title : undefined };
};

export const loadFile = async data => {
  const [file] = await files.get({ _id: data._id });
  return { ...data, file, title: file.originalname ? file.originalname : undefined };
};

export const extraTemplate = data =>
  `of type ${
    data.templateData
      ? data.templateData.name
      : `(${data.template ? data.template.toString() : 'unassigned'})`
  }`;

export const extraAttachmentLanguage = data =>
  data.entity
    ? `of entity '${data.entity.title}' (${data.entity.sharedId}) ${formatLanguage(
        data.entity.language
      )} version`
    : null;

export const searchName = data =>
  data.search ? `${data.search.searchTerm} (${data.searchId})` : data.searchId;

export const updatedFile = data => {
  let name;
  if (data.toc) {
    name = 'ToC, ';
  } else {
    name = data.pdfinfo ? 'Pdf info, ' : '';
  }
  return `${name}${data.title}`;
};

export const groupMembers = data => {
  const members = data.members.map(member => member.username).join(', ');
  return members.length > 0 ? `with members: ${members}` : 'with no members';
};
