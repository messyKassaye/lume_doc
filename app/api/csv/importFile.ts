import fs from 'fs';
import path from 'path';
import { Readable } from 'stream';

import { generateFileName, fileFromReadStream, uploadsPath } from 'api/files/filesystem';
import { createError } from 'api/utils';
import zipFile from 'api/utils/zipFile';

const extractFromZip = async (zipPath: string, fileName: string) => {
  const readStream = await zipFile(zipPath).findReadStream(entry => entry === fileName);

  if (!readStream) {
    throw createError(`${fileName} file not found`);
  }

  return readStream;
};

export class ImportFile {
  filePath: string | Readable;

  constructor(filePath: string | Readable) {
    this.filePath = filePath;
  }

  async readStream(fileName = 'import.csv') {
    if (this.filePath instanceof Readable) {
      return this.filePath;
    }
    if (path.extname(this.filePath) === '.zip') {
      return extractFromZip(this.filePath, fileName);
    }
    return fs.createReadStream(this.filePath);
  }

  async extractFile(fileName: string, destination: string | undefined = undefined) {
    const generatedName = generateFileName({ originalname: fileName });

    await fileFromReadStream(generatedName, await this.readStream(fileName), destination);

    return {
      destination: destination || uploadsPath(),
      originalname: fileName,
      filename: generatedName,
    };
  }
}

const importFile = (filePath: string | Readable) => new ImportFile(filePath);

export default importFile;
