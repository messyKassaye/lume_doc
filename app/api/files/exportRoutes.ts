import { Application, Request, Response, NextFunction } from 'express';
import { createWriteStream, unlink } from 'fs';
import errorLog from 'api/log/errorLog';
import { search } from 'api/search';
import { validation } from '../utils';
import { isArray } from 'util';
import { temporalFilesPath, generateFileName } from './filesystem';
import { CSVExporter } from 'api/csv';
import settings from 'api/settings';

export default (app: Application) => {
  const parseQueryProperty = (query: any, property: string) =>
    query[property] ? JSON.parse(query[property]) : query[property];

  const generateExportFileName = (databaseName: string = '') =>
    `${databaseName}-${new Date().toISOString()}.csv`;

  const removeTempFile = (filePath: string) => () => {
    unlink(filePath, err => {
      if (err) errorLog.error(`Error unlinking exported file: ${filePath}`);
    });
  };

  app.get(
    '/api/export',
    validation.validateRequest({
      properties: {
        query: {
          properties: {
            filters: { type: 'string' },
            types: { type: 'string' },
            allAggregations: { type: 'string' },
            userSelectedSorting: { type: 'string' },
            order: { type: 'string' },
            sort: { type: 'string' },
            limit: { type: 'string' },
            searchTerm: { type: 'string' },
            includeUnpublished: { type: 'string' },
            unpublished: { type: 'string' },
            ids: { type: 'string' },
          },
        },
      },
    }),
    async (req: Request, res: Response, next: NextFunction) => {
      const temporalFilePath = temporalFilesPath(generateFileName({ originalname: 'export.csv' }));
      try {
        req.query.filters = parseQueryProperty(req.query, 'filters');
        req.query.types = parseQueryProperty(req.query, 'types');
        req.query.unpublished = parseQueryProperty(req.query, 'unpublished');
        req.query.includeUnpublished = parseQueryProperty(req.query, 'includeUnpublished');
        req.query.allAggregations = parseQueryProperty(req.query, 'allAggregations');

        req.query.ids = parseQueryProperty(req.query, 'ids');
        if (!isArray(req.query.ids)) delete req.query.ids;

        const results = await search.search(req.query, req.language, req.user);
        // eslint-disable-next-line camelcase
        const { dateFormat = '', site_name } = await settings.get();

        const exporter = new CSVExporter();

        const fileStream = createWriteStream(temporalFilePath);
        const exporterOptions = { dateFormat, language: req.language };

        await exporter.export(results, req.query.types, fileStream, exporterOptions);

        res.download(
          temporalFilePath,
          generateExportFileName(site_name),
          removeTempFile(temporalFilePath)
        );
      } catch (e) {
        removeTempFile(temporalFilePath)();
        next(e);
      }
    }
  );
};
