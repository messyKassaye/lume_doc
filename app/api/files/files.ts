import { deleteUploadedFiles } from 'api/files/filesystem';
import connections from 'api/relationships';
import { search } from 'api/search';
import entities from 'api/entities';
import request from 'shared/JSONRequest';
import model from './filesModel';
import { validateFile } from '../../shared/types/fileSchema';
import { FileType } from '../../shared/types/fileType';

export const files = {
  async save(_file: FileType) {
    const file = { ..._file };
    if (file.url && !file._id) {
      const response = await request.head(file.url);
      const mimetype = response.headers.get('content-type') || undefined;
      file.mimetype = mimetype;
    }

    const savedFile = await model.save(await validateFile(file));
    await search.indexEntities({ sharedId: savedFile.entity }, '+fullText');
    return savedFile;
  },

  get: model.get.bind(model),

  async delete(query: any = {}) {
    const toDeleteFiles: FileType[] = await model.get(query);
    await model.delete(query);
    if (toDeleteFiles.length > 0) {
      await connections.delete({ file: { $in: toDeleteFiles.map(f => f._id?.toString()) } });
      await deleteUploadedFiles(toDeleteFiles);
      await search.indexEntities(
        { sharedId: { $in: toDeleteFiles.map(f => f.entity?.toString()) } },
        '+fullText'
      );
    }

    return toDeleteFiles;
  },

  async tocReviewed(_id: string, language: string) {
    const savedFile = await files.save({ _id, generatedToc: false });
    const sameEntityFiles = await files.get({ entity: savedFile.entity }, { generatedToc: 1 });
    const [entity] = await entities.get({
      sharedId: savedFile.entity,
    });

    await entities.save(
      {
        _id: entity._id,
        sharedId: entity.sharedId,
        template: entity.template,
        generatedToc: sameEntityFiles.reduce<boolean>(
          (generated, file) => generated || Boolean(file.generatedToc),
          false
        ),
      },
      { user: {}, language }
    );

    return savedFile;
  },
};
