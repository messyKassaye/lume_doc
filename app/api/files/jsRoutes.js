import fs from 'fs';
import proxy from 'express-http-proxy';

import entities from 'api/entities';
import mailer from 'api/utils/mailer';
import { search } from 'api/search';
import settings from 'api/settings';
import { processDocument } from 'api/files/processDocument';
import { files } from 'api/files/files';
import { uploadsPath, attachmentsPath, generateFileName } from 'api/files/filesystem';
import cors from 'cors';
import activitylogMiddleware from 'api/activitylog/activitylogMiddleware';
import { validation, createError } from '../utils';
import { captchaAuthorization } from '../auth';

import { uploadMiddleware } from './uploadMiddleware';

const storeFile = (pathFunction, file) =>
  new Promise((resolve, reject) => {
    const filename = generateFileName(file);
    fs.appendFile(pathFunction(filename), file.buffer, err => {
      if (err) {
        reject(err);
      }
      resolve(Object.assign(file, { filename, destination: pathFunction() }));
    });
  });

const routes = app => {
  const corsOptions = {
    origin: true,
    methods: 'POST',
    credentials: true,
    optionsSuccessStatus: 200,
  };

  app.options('/api/public', cors(corsOptions));
  app.post(
    '/api/public',
    cors(corsOptions),
    uploadMiddleware.multiple(),
    captchaAuthorization(),
    activitylogMiddleware,
    (req, _res, next) => {
      try {
        req.body.entity = JSON.parse(req.body.entity);
        if (req.body.email) {
          req.body.email = JSON.parse(req.body.email);
        }
      } catch (err) {
        next(err);
        return;
      }
      next();
    },
    validation.validateRequest({
      properties: {
        body: {
          properties: {
            email: {
              properties: {
                to: { type: 'string' },
                from: { type: 'string' },
                text: { type: 'string' },
                html: { type: 'string' },
                subject: { type: 'string' },
              },
              required: ['to', 'from', 'text', 'subject'],
            },
          },
        },
      },
    }),
    async (req, res, next) => {
      const { allowedPublicTemplates } = await settings.get();
      const { entity, email } = req.body;

      if (!allowedPublicTemplates || !allowedPublicTemplates.includes(entity.template)) {
        next(createError('Unauthorized public template', 403));
        return;
      }

      const newEntity = await entities.save(entity, { user: {}, language: req.language });

      const attachments = [];
      if (req.files.length) {
        await Promise.all(
          req.files
            .filter(file => file.fieldname.includes('attachment'))
            .map(file =>
              storeFile(attachmentsPath, file).then(_file =>
                attachments.push({
                  ..._file,
                  entity: newEntity.sharedId,
                  type: 'attachment',
                })
              )
            )
        );
      }

      await Promise.all(attachments.map(attachment => files.save(attachment)));

      const file = req.files.find(_file => _file.fieldname.includes('file'));
      if (file) {
        storeFile(uploadsPath, file).then(async _file => {
          await processDocument(newEntity.sharedId, _file);
          await search.indexEntities({ sharedId: newEntity.sharedId }, '+fullText');
          req.emitToSessionSocket('documentProcessed', newEntity.sharedId);
        });
      }

      if (email) {
        await mailer.send(email);
      }

      res.json(newEntity);
    }
  );

  app.post('/api/remotepublic', async (req, res, next) => {
    const { publicFormDestination } = await settings.get({}, { publicFormDestination: 1 });
    proxy(publicFormDestination, {
      limit: '20mb',
      proxyReqPathResolver() {
        return '/api/public';
      },
      proxyReqOptDecorator(proxyReqOpts) {
        const { tenant, ...headers } = proxyReqOpts.headers;
        return {
          ...proxyReqOpts,
          headers: { ...headers },
        };
      },
    })(req, res, next);
  });
};

export default routes;
export { routes };
