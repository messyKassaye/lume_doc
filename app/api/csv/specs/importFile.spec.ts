import path from 'path';
import fs from 'fs';
import { streamToString } from 'api/files/filesystem';

import importFile from '../importFile';
import { createTestingZip } from './helpers';

describe('importFile', () => {
  beforeAll(async () => {
    await createTestingZip(
      [path.join(__dirname, '/zipData/import.csv'), path.join(__dirname, '/zipData/file1.txt')],
      'ImportFile.zip'
    );

    await createTestingZip([path.join(__dirname, '/zipData/file1.txt')], 'badFile.zip');
  });

  afterAll(() => {
    fs.unlinkSync(path.join(__dirname, '/zipData/ImportFile.zip'));
    fs.unlinkSync(path.join(__dirname, '/zipData/badFile.zip'));
  });

  describe('readStream', () => {
    it('should return a readable stream for the csv file', async () => {
      const file = importFile(path.join(__dirname, '/test.csv'));
      const fileContents = await streamToString(await file.readStream());
      expect(fileContents).toMatchSnapshot();
    });

    describe('when file is a a zip', () => {
      it('should return a stream for a file that should be called import.csv by default', async () => {
        const file = importFile(path.join(__dirname, '/zipData/ImportFile.zip'));
        const fileContents = await streamToString(await file.readStream());
        expect(fileContents).toMatchSnapshot();
      });

      describe('when passing a filename', () => {
        it('should return a read stream for that file', async () => {
          const file = importFile(path.join(__dirname, '/zipData/ImportFile.zip'));
          const fileContents = await streamToString(await file.readStream('file1.txt'));
          expect(fileContents).toMatchSnapshot();
        });
      });
    });
  });

  describe('when there is no import.csv on the zip', () => {
    it('should throw an error', async () => {
      const file = importFile(path.join(__dirname, '/zipData/badFile.zip'));
      let error;
      try {
        await file.readStream();
      } catch (e) {
        error = e;
      }

      expect(error.message.match(/import\.csv/)).not.toBeNull();
    });
  });
});
