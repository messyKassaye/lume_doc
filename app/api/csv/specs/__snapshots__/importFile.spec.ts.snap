// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`importFile readStream should return a readable stream for the csv file 1`] = `
"Title , text label  , numeric label, non configured, select label, not defined type, geolocation_geolocation

title1, text value 1, 1977, ______________, thesauri1   , notType1                 , 1|1,
title2, text value 2, 2019, ______________, thesauri2   , notType2                 ,    ,
title3, text value 3, 2020, ______________, thesauri2   , notType3                 , 0|0,
"
`;

exports[`importFile readStream when file is a a zip should return a stream for a file that should be called import.csv by default 1`] = `
"Title , file , attachments,

file1, 1.pdf, att1.doc,
file2, 2.pdf, att1.doc|att2.jpg,
file3, 3.pdf,                  ,
"
`;

exports[`importFile readStream when file is a a zip when passing a filename should return a read stream for that file 1`] = `
"file1.txt
"
`;
